### DISCUSSION TITLE
Visual sub-tree to Bitmap for animations

### DISCUSSION THREAD

**Discussion by llfab** [AvaloniaUI/Avalonia#9319] [Q&A] [answered]
I just noticed that I can render any ```IVisual``` into a Bitmap. 

For some fancy animations with complex controls inside you may want to create such a still Bitmap for a visual sub-tree, show that Bitmap instead of the complex sub-tree, start your fancy (transitioning) animation and once the animation has reached its final position, show the original sub-tree again. 

Why that? Because your complex sub-tree may not render smoothly during the animation, but the Bitmap will.

So, is that the proposed way of doing such a thing?

**Comment by maxkatz6** [comment, marked as answer]
It's definitely a way to do thinks. Although with potentially better performance you will get more complex code.

Avalonia 11.0 has new feature of Composition API, which allows to animate render-visuals on the render thread, avoiding layout recalculations and interaction with the UI thread at all.
We don't have many examples yet (planned to be shared after the release) nor documentation, but in overall API follows the same [what UWP has](https://xamlbrewer.wordpress.com/2016/01/04/using-the-composition-api-in-uwp-apps/).

You also can find this simple implicit animation demo https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog/Pages/CompositionPage.axaml.cs#L84

**Reply by maxkatz6** [reply]
From my experience with older versions, I also used your approach of rendering into a bitmap first. It works well, but there are couple of problems I faced:
1. All the content is clipped by the bounds of the bitmap. When original control might have content rendered outside of its bounds (shadows, or not clipped children).
2. If your control is inside of the Viewbox, your control will be a less quality, as it's not a "vector", but "raster" now.

**Reply by llfab** [reply]
Thanks for the insights! You may know that we plan to integrate our vulkan based 3D renderer into Avalonia on Vulkan. In a previous product we neded to animate (move, resize) that around. In Qt/Qml we used ShaderEffectSource because such 3D content controls sometimes don't do well on resize. Low quality during animation seems less of a problem. And yes: the code for this definitely is more complex.

**Comment by kekekeks** [comment]
That can be achieved by implementing `CacheMode` / `BitmapCache`, which is planned but without ETA.

**Reply by llfab** [reply]
Very nice!

**Reply by h4lfheart** [reply]
Any update on this? Working with a port of a WPF library(https://github.com/BAndysc/nodify-avalonia) and it's missing this CacheMode which would help performance a lot!!
